iroh-bytes/src/store/file.rs

```diff
@@ -0,0 +1,2483 @@
+        Ok(tokio::task::spawn_blocking(move || this.import_bytes_sync(data, format)).await??)
+    }
+
+    async fn import_stream(
+        &self,
+        mut data: impl Stream<Item = io::Result<Bytes>> + Unpin + Send + 'static,
+        format: BlobFormat,
+        progress: impl ProgressSender<Msg = ImportProgress> + IdGenerator,
+    ) -> io::Result<(TempTag, u64)> {
+        let this = self.clone();
+        let id = progress.new_id();
+        // write to a temp file
+        let temp_data_path = this.0.temp_file_name();
+        let name = temp_data_path
+            .file_name()
+            .expect("just created")
+            .to_string_lossy()
+            .to_string();
+        progress.send(ImportProgress::Found { id, name }).await?;
+        let mut writer = tokio::fs::File::create(&temp_data_path).await?;
+        let mut offset = 0;
+        while let Some(chunk) = data.next().await {
+            let chunk = chunk?;
+            writer.write_all(&chunk).await?;
+            offset += chunk.len() as u64;
+            progress.try_send(ImportProgress::CopyProgress { id, offset })?;
+        }
+        writer.flush().await?;
+        drop(writer);
+        let file = ImportSource::TempFile(temp_data_path);
+        Ok(tokio::task::spawn_blocking(move || {
+            this.0.finalize_import_sync(file, format, id, progress)
+        })
+        .await??)
+    }
+
+    async fn set_tag(&self, name: Tag, hash: Option<HashAndFormat>) -> io::Result<()> {
+        Ok(self.0.set_tag(name, hash).await?)
+    }
+
+    async fn create_tag(&self, hash: HashAndFormat) -> io::Result<Tag> {
+        Ok(self.0.create_tag(hash).await?)
+    }
+
+    async fn delete(&self, hashes: Vec<Hash>) -> io::Result<()> {
+        Ok(self.0.delete(hashes).await?)
+    }
+
+    async fn gc_start(&self) -> io::Result<()> {
+        self.0.clear_protected().await?;
+        Ok(())
+    }
+
+    fn temp_tag(&self, value: HashAndFormat) -> TempTag {
+        self.0.temp_tag(value)
+    }
+}
+
+impl Actor {
+    fn new(
+        path: &Path,
+        options: Options,
+        temp: Arc<RwLock<TempCounterMap>>,
+        rt: tokio::runtime::Handle,
+    ) -> ActorResult<(Self, flume::Sender<ActorMessage>)> {
+        let db = redb::Database::create(path)?;
+        let txn = db.begin_write()?;
+        let tables = Tables::new(&txn)?;
+        drop(tables);
+        txn.commit()?;
+        let (tx, rx) = flume::unbounded();
+        let tx2 = tx.clone();
+        let tx3 = tx.clone();
+        let on_file_create: CreateCb = Arc::new(move |hash| {
+            // todo: make the callback allow async
+            tx2.send(ActorMessage::OnMemSizeExceeded { hash: *hash })
+                .ok();
+            Ok(())
+        });
+        let on_drop: DropCb = Arc::new(move |hash, refcount| {
+            // the count includes the one that is being dropped, so we check for 2.
+            // it will be 1 when the last one is being dropped.
+            if refcount == 2 {
```

Comment:
wouldn't this still be `1`, and the count is `0` after the drop?

Reply:
We don't want the last drop but the second to last drop. We know that 1 entry is in the entry map, so we want to do the cb if all *other* clones are done.